fix: client retries timeout/blockhash issue

amilz opened the pull request:
Fixes #4 

### Issue
Client retries uses default 15 second retry time-- when using `finalized` commitment, retry can attempt before the transaction has been confirmed even though it's landed (and just in confirmed on the network). This causes client errors showing failure despite the transaction having landed.

### Fix
- Allow for user to override default `timeout` value
- Improved defaults for `timeout` depending on commitment
- Added tests for various commitment/retry levels
- Adds warning for using client retry with `finalized` (b/c this will likely result in stale blockhashes)
- Move our `getLatestBlockhash` call to the end of the function to ensure it's as fresh as possible before sending to the network

### Other 
- Upgraded @solana-program/compute-budget to v0.9.0 (and all peer deps to ensure ci pass)
- Added @solana/promises to dependencies
- Fix some `any` use

amilz committed:
fix: client retries timeout/blockchash issue

- Upgraded @solana-program/compute-budget to v0.9.0
- added @solana/promises to dependencies.
- Updated core transaction and connection logic in src/lib,
- added/modified tests for connect and transactions modules to improve coverage.

amilz committed:
chore: update versions for ci tests

amilz committed:
chore: use npm

amilz committed:
chore: fix test name

mikemaccana commented on src/tests/transactions.test.ts:
```diff
@@ -0,0 +1,134 @@
+import { connect, Connection } from "..";
+import { isSolanaError, KeyPairSigner, lamports, SOLANA_ERROR__JSON_RPC__SERVER_ERROR_SEND_TRANSACTION_PREFLIGHT_FAILURE, SOLANA_ERROR__TRANSACTION_ERROR__ALREADY_PROCESSED, SolanaError } from "@solana/kit";
+import { getTransferSolInstruction } from "@solana-program/system";
+import { SOL } from "../lib/constants";
```
This file needs formatting!

amilz replied:
added a prettier script and ran it!

mikemaccana commented on src/lib/connect.ts:
```diff
@@ -273,6 +292,7 @@ export const connect = (
     getTokenAccountBalance,
     getPDAAndBump,
     checkTokenAccountIsClosed,
+    getTokenMetadata,
```
Thanks! Looks l;ike your branch is out of date, can please update it?

mikemaccana commented on src/lib/connect.ts:
```diff
@@ -204,12 +210,23 @@ export const connect = (
+  
+  switch (true) {
+    case clusterNameOrURL.includes("mainnet"):
+      sendAndConfirmTransaction = sendAndConfirmTransactionFactory({ rpc, rpcSubscriptions: typedRpcSubscriptions as RpcSubscriptionsMainnet<SolanaRpcSubscriptionsApi> });
```
Thanks I've been meaning to fix this for a while. 😃 Would:


```
rpcSubscriptions: RpcSubscriptionsMainnet<SolanaRpcSubscriptionsApi>
```

work? I don't understand why we need the type coersion.

amilz replied:
I used a generic. I think if we use Mainnet it'll cause issues w/ the airdrop polling. Not 100% sure though--feel free to change:

```ts
  let rpcSubscriptions: RpcSubscriptions<SolanaRpcSubscriptionsApi>;
```

mikemaccana commented on src/lib/smart-transactions.ts:
```diff
@@ -127,10 +133,30 @@ export const sendTransactionWithRetries = async (
+      break;
+    default:
+      timeout = DEFAULT_TRANSACTION_TIMEOUT;
+      break;
```
Good thinking! Make sure you update CHANGELOG.md with the changes (make an '*in progress* heading for the next release if you need to).

mikemaccana reviewed (changes requested):
Awesome, thankyou! Just need the the branch updated, a CHANGELOG with the additions and some minor but but otherwise good to go. Thanks @amilz ! 🙏

amilz committed:
Merge remote-tracking branch 'upstream/main' into fix-transaction-lifetime

amilz committed:
prettier

- add prettier script to package.json
- run prettier

amilz committed:
chore: cleanup rpcSubscriptions types

amilz committed:
chore: changelog

amilz committed:
chore: remove unused

amilz committed:
chore: fix failing test

getTransaction doesn't support processed, so we need to fallback to confirmed if we get to the getTransaction step
https://solana.com/docs/rpc/http/gettransaction

amilz committed:
chore: try longer timeout for ci

CI environment is slow/rate-limited on public endpoints. Going to try higher timeout for CI.

amilz committed:
chore: catch simulation error

amilz committed:
chore: remove no longer relevant test

this test no longer makes sense given that we handle simulation error in our retry logic

amilz commented:
@mikemaccana for some reason the CI did not like those cleanups. I think the local validator can run kind of slow on the GH action--so caught a fringe issue in error handling in `smart-transaction.ts` while loop. Make sure you're good with it. Also had to add prettier script -- ran it, so there are a few additional edited files in this PR than before--just costmetic. Once this one merges, I'll update the other open prs.

mikemaccana commented:
@amilz I ran prettier and a made a formatting-only commit, then merged #5 and #6 - can you please update this one too and I'll get it in.

amilz committed:
Merge remote-tracking branch 'upstream/main' into fix-transaction-lifetime

amilz commented:
@mikemaccana i think rdy

amilz commented:
Resolved in #10
